PL2/PL2-core/src/main/java/nl/tudelft/pl2016gr2/core/algorithms/SplitGraphs.java

```diff
@@ -0,0 +1,151 @@
+ * by the original graph.</li>
+ * </ul>
+ * </p>
+ *
+ * @author Wouter Smit
+ */
+public class SplitGraphs {
+
+  @TestId(id = "graph_field")
+  private final OriginalGraph mainGraph;
+
+  /**
+   * Instantiates an algorithmic class on a <code>OriginalGraph</code>.
+   * <p>
+   * When {@link #getSubgraph(Collection)} is called, the class will construct a new instance of
+   * the
+   * graph containing only the specified genomes.
+   * </p>
+   *
+   * @param mainGraph The graph to split
+   */
+  public SplitGraphs(OriginalGraph mainGraph) {
+    this.mainGraph = mainGraph;
+  }
+
+
+  /**
+   * Builds a subgraph of the {@link #mainGraph} containing only the nodes that contain (at least)
+   * one of the specified genomes.
+   *
+   * @param genomes The genomes that should be included in the resulting graph
+   * @return An <code>OriginalGraph</code> that is a subgraph of <code>mainGraph</code>
+   * @throws NoSuchElementException Iff the genome collection is not a
+   *                                subset of the main graph's genomes.
+   */
+  public OriginalGraph getSubgraph(Collection<String> genomes) {
+    if (!mainGraph.getGenoms().containsAll(genomes)) {
+      throw new NoSuchElementException("All genomes must be present in the main graph.");
+    }
+    OriginalGraph subGraph = buildSubgraph(genomes);
+    return prune(subGraph, genomes);
+  }
+
+  /**
+   * Creates a new graph containing only the nodes that will be part of the subGraph.
+   * <p>
+   * The nodes are not a deep copy, however the graph object is newly instantiated.
+   * </p>
+   *
+   * @param genomes The genomes that should be included in the subgraph
+   * @return The subgraph
+   */
+  private OriginalGraph buildSubgraph(Collection<String> genomes) {
+    OriginalGraph subGraph = new OriginalGraph();
+    // Add all nodes for every genome
+    for (String genome : genomes) {
+      Iterator<Node> iterator = mainGraph.iterator(genome);
+      // add genome root
+      subGraph.addNode(iterator.next());
+
+      while (iterator.hasNext()) {
+        subGraph.addNode(iterator.next());
+      }
+
+    }
+    return subGraph;
+  }
+
+
+  /**
+   * Removes all redundant links from the nodes in the graph.
+   * <p>
+   * Redundant links are links that point to a {@link Node} that is not part of the subgraph.
+   * The method will return a new {@link OriginalGraph} object with new <code>Node</code>s.
+   * The Nodes will contain identical fields, except for the {@link Node#genomes}, {@link
+   * Node#inLinks} and {@link Node#outLinks}.
+   * </p>
+   * <p>
+   * While this method is intended solely for subgraphs, it will provide the same behaviour when
+   * called with any graph.
+   * </p>
+   *
+   * @param graph The graph to prune.
+   * @return A new <code>OriginalGraph</code> with only the relevant links.
+   */
+  private OriginalGraph prune(OriginalGraph graph, Collection<String> genomes) {
```

Comment:
If you change the class in such a way that you create new nodes in the buildsubgraph, you should edit this method in such a way that you still get the inlinks and outlinks of the nodes of the original graph.